File: src/common/py/data/entities/project/features/project_feature.py
import abc
import typing
from dataclasses import dataclass, field

from dataclasses_json import dataclass_json

from ...properties import ProfileID

ProjectFeatureID = str


@dataclass_json
@dataclass
class ProjectFeature(abc.ABC):
    """
    Base data class for a project feature.
    """

    enabled_metadata_profiles: typing.List[ProfileID] = field(default_factory=list)

File: src/common/web/data/entities/project/features/ProjectFeature.ts
import { Type } from "class-transformer";

import { type ProfileID } from "../../../../ui/components/propertyeditor/PropertyProfile";

/**
 * The project feature identifier type.
 */
export type ProjectFeatureID = string;

/**
 * Base data class for a project feature.
 */
export abstract class ProjectFeature {
    // @ts-ignore
    @Type(() => [String, String])
    public readonly enabled_metadata_profiles: ProfileID[] = [];

    /**
     * The ID of this feature.
     */
    public abstract get featureID(): ProjectFeatureID;

    public constructor(enabledProfiles: ProfileID[] = []) {
        this.enabled_metadata_profiles = enabledProfiles;
    }
}
